### PR TITLE
FaaS positioning & supporting user-defined methods to handle more HTTP methods

### LINKED ISSUE #2063: Support user-defined functions to handle more HTTP methods for realtime APIs
### Implementation details

We could add an additional function to the Predictor API (e.g. `get()` or `handle_get()`), which would be called when a GET request is sent to an API endpoint.

* Check whether we enforce for the presence of the constructor for the task api.

## Questions

* Which HTTP methods should be supported? (GET/PUT/DELETE etc.)
  * All RESTful verbs. POST/GET/PUT/PATCH/DELETE.
* What is the naming scheme of the methods?
  * "handle_<verb-name>".
* What about the name of the method for gRPC?
  * "handle"/"run". We can go with **handle**, but if it's easy, we should allow for the user to specify any number of service methods (time-capped@.25).
* How do we still name the methods for Batch/Async?
  * Short-path names: "handle_batch"/"handle_async". 
  * Long-path names: from cortex.handler import Realtime. And Realtime class (like others) will enforce some methods to be defined. Realtime is an abstract class with default implementations for each verb method.

* If / how would they affect autoscaling? Assuming the user does inexpensive tasks within `get()`, this could be ignored for autoscaling purposes. Otherwise, it should be counted as an in-flight request.
  * We can either count it as an in-flight request for simplicity reasons or we can enable decorators for the user to toggle the counting or not.
  * There's a risk in making GET countable that will lead to unreasonable scale-ups. This can be caused by the user's constant API gets (using the CLI/Python client).
  * **For now, everything is in the queue and everything is counted as in-flight requests.**
* Should the GET request bypass the request queue, and immediately execute in an API thread?
  * No. For simplicity reasons, let's use the same thread pool for all verbs.
* If / how should this be recorded in the metrics, and shown on the metrics dashboard?
  * In Grafana we would still show the same metrics, but only for the main verb (POST). Not sure what we can do about gRPC in this case.
  * We could be counting the requests (in cortex get <api-name>) by verb name. We'd need to add the verb as another metric dimension.
  * We can add another field in the "predictor" section (like "tracked_verb") to indicate which verb needs to be tracked (in Grafana and cortex get). A corollary is that counting the in-flight requests for all other verbs no longer seems necessary.
* How will the latency reporting be affected in cortex get?  
  * Latencies would be grouped by verb name.
  * Or do nothing if we add the "tracked_verb" field.
  * **Let's aggregate them instead (aka do nothing).**

### PR DESCRIPTION
**Pull request by RobertLucian**
Closes #2063.

---

##### Methods for Realtime API

1. HTTP:
  * `__init__`, `handle_<verb>`, conditional `load_model`.
  * Remove “post_predict” in favour of returning a callback instead.
  * Dynamic-batcher for the post verb (`handle_post` method) only.
1. gRPC:
  * `__init__`,  any number of methods as indicated by the protobuf definition (RPC methods), conditional `load_model`.

##### Methods for Batch API

* `__init__`, `handle_batch`, `handle_on_job_complete`.

##### Methods for Async API

* `__init__`, `handle_async`.

##### Name of the predictor clients

Renamed `python_client`/`tensorflow_client` to `model_client`/`tensorflow_client`.

##### Changes to the API spec

Predictor section renamed to `handler`.

---

checklist:

- [x] run `make test` and `make lint`
- [x] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)
- [x] run e2e
- [x] update examples
- [x] update docs and add any new files to `summary.md` (view in [gitbook](https://cortex-labs.gitbook.io/staging/-MOmCGMADSRNQahK3Kox/) after merging)

### CODE REVIEW

File: pkg/operator/resources/resources.go [hunk 1]
```diff
@@ -598,6 +606,23 @@ func GetAPI(apiName string) ([]schema.APIResponse, error) {
 		}
 	}
 
+	// best effort
+	if config.K8s != nil && apiResponse[0].Spec.Kind == userconfig.RealtimeAPIKind && !apiResponse[0].Spec.Handler.IsGRPC() && (apiResponse[0].Spec.Handler.MultiModelReloading != nil || apiResponse[0].Spec.Handler.Models != nil) {
+		internalAPIEndpoint, err := config.K8s.InternalServiceEndpoint("api-"+apiResponse[0].Spec.Name, _defaultAPIPortInt32)
+		if err != nil {
+			operatorLogger.Warn(errors.Wrap(err, fmt.Sprintf("api %s", apiResponse[0].Spec.Name)))
```
Comment (RobertLucian) [inline pkg/operator/resources/resources.go:613]: return `return apiResponse, nil` where appropriate.

File: test/apis/pytorch/iris-classifier/predictor.py [hunk 2]
```diff
@@ -41,3 +42,14 @@ def predict(self, payload):
 
         # Translate the model output to the corresponding label string
         return labels[torch.argmax(output[0])]
+
+    def handle_get(self):
```
Comment (RobertLucian) [inline test/apis/pytorch/iris-classifier/predictor.py:46]: Remove `handle_get` and `handle_put` methods.

File: pkg/cortex/serve/cortex_internal/lib/api/batch.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            "optional_args": ["job_spec", "metrics_client"],
+        },
+        {
+            "name": "handle_batch",
+            "required_args": ["self"],
+            "optional_args": ["payload", "batch_id"],
+        },
+    ],
+    "optional": [
+        {"name": "handle_on_job_complete", "required_args": ["self"]},
```
Comment (miguelvr) [inline pkg/cortex/serve/cortex_internal/lib/api/batch.py:54]: `handle_on_job_complete` sounds too cumbersome, what do you think of keeping the old name?

File: pkg/cortex/serve/cortex_internal/serve/serve.py [hunk 4]
```diff
@@ -203,57 +220,39 @@ def predict(request: Request):
...
+    if "query_params" in local_cache["handle_fn_args"][verb]:
         kwargs["query_params"] = request.query_params
-    if "response" in local_cache["post_predict_fn_args"]:
-        kwargs["response"] = response
 
     return kwargs
 
 
 def get_summary():
-    response = {"message": API_SUMMARY_MESSAGE}
```
Comment (deliahu) [inline pkg/cortex/serve/cortex_internal/serve/serve.py:244]: initialize an empty response

File: pkg/cortex/serve/cortex_internal/serve/serve.py [hunk 5]
```diff
@@ -309,33 +312,45 @@ def check_if_crons_have_failed():
...
+
+        local_cache["handle_fn_args"] = {}
+        for verb in ["post", "get", "put", "patch", "delete"]:
+            if util.has_method(handler_impl, f"handle_{verb}"):
+                local_cache["handle_fn_args"][verb] = inspect.getfullargspec(
+                    getattr(handler_impl, f"handle_{verb}")
+                ).args
+        if len(local_cache["handle_fn_args"]) == 0:
+            raise UserRuntimeException(
+                "no user-defined handle_<verb> method found in handler class; define at least one verb handler (post, get, put, patch, delete)"
```
Comment (deliahu) [inline pkg/cortex/serve/cortex_internal/serve/serve.py:325]: nit:

```
(`handle_post`, `handle_get`, `handle_put`, `handle_patch`, `handle_delete`)
```

File: pkg/cortex/serve/cortex_internal/serve/serve.py [hunk 6]
```diff
@@ -309,33 +312,45 @@ def check_if_crons_have_failed():
...
-                batch_interval=dynamic_batching_config["batch_interval"]
-                / NANOSECONDS_IN_SECOND,  # convert nanoseconds to seconds
-            )
+            dynamic_batching_config = api.api_spec["handler"]["server_side_batching"]
 
-        if util.has_method(predictor_impl, "post_predict"):
-            local_cache["post_predict_fn_args"] = inspect.getfullargspec(
-                predictor_impl.post_predict
-            ).args
+            if "post" in local_cache["handle_fn_args"]:
```
Comment (deliahu) [inline pkg/cortex/serve/cortex_internal/serve/serve.py:331]: nit: return error if handle_post is not defined

File: pkg/cortex/serve/nginx.conf.j2 [hunk 7]
```diff
@@ -113,36 +113,44 @@ http {
...
+
         location / {
             deny all;
         }
 
-        location ~ ^/(predict/?|)$ {
+        location = / {
             # CORS (inspired by https://enable-cors.org/server_nginx.html)
+
+            # for all CRUD methods
```
Comment (deliahu) [inline pkg/cortex/serve/nginx.conf.j2:137]: test CORS

File: pkg/lib/configreader/string.go [hunk 8]
```diff
@@ -277,6 +279,26 @@ func ValidateStringVal(val string, v *StringValidation) error {
 		}
 	}
 
+	matchedPrefixes := 0
+	for _, allowedPrefix := range v.AllowedPrefixes {
```
Comment (deliahu) [inline pkg/lib/configreader/string.go:283]: move `if len(v.AllowedPrefixes) > 0` to the top

File: pkg/lib/configreader/string.go [hunk 9]
```diff
@@ -43,6 +43,8 @@ type StringValidation struct {
 	Prefix                               string
 	Suffix                               string
 	InvalidPrefixes                      []string
+	AllowedPrefixes                      []string
+	AllowedSuffixes                      []string
```
Comment (deliahu) [inline pkg/lib/configreader/string.go:47]: duplicate this in `string_ptr.go`

File: pkg/operator/operator/k8s.go [hunk 10]
```diff
@@ -116,10 +116,10 @@ func TaskInitContainer(api *spec.API) kcore.Container {
 func InitContainer(api *spec.API) kcore.Container {
 	downloadArgs := ""
 
-	switch api.Predictor.Type {
-	case userconfig.TensorFlowPredictorType:
+	switch api.Handler.Type {
+	case userconfig.TensorHandlerType:
```
Comment (deliahu) [inline pkg/operator/operator/k8s.go:120]: nit: `TensorFlowHandlerType`

File: pkg/lib/k8s/service.go [hunk 11]
```diff
@@ -156,6 +157,19 @@ func (c *Client) ListServicesWithLabelKeys(labelKeys ...string) ([]kcore.Service
...
 
+func (c *Client) InternalServiceEndpoint(serviceName string, portNumber int32) (string, error) {
+	service, err := c.GetService(serviceName)
+	if err != nil {
+		return "", err
+	}
+
+	if service == nil {
+		return "", ErrorServiceNotFound(serviceName)
+	}
```
Comment (deliahu) [inline pkg/lib/k8s/service.go:168]: Remove this check

File: pkg/lib/k8s/service.go [hunk 12]
```diff
@@ -156,6 +157,19 @@ func (c *Client) ListServicesWithLabelKeys(labelKeys ...string) ([]kcore.Service
...
+
+	return fmt.Sprintf("http://%s.%s.svc.cluster.local:%d", serviceName, service.Namespace, portNumber), nil
```
Comment (deliahu) [inline pkg/lib/k8s/service.go:170]: `c.Namespace`

File: pkg/operator/schema/schema.go [hunk 13]
```diff
@@ -85,8 +86,12 @@ type ErrorResponse struct {
 	Message string `json:"message"`
 }
 
+type RealtimeModelMetadata struct {
+	APITFModelSummary     *APITFLiveReloadingSummary `json:"tf_model_summary"`
```
Comment (deliahu) [inline pkg/operator/schema/schema.go:90]: nit: `API` prefix can be removed

File: test/apis/pytorch/iris-classifier/predictor.py [hunk 14]
```diff
@@ -5,11 +5,12 @@
 from botocore import UNSIGNED
 from botocore.client import Config
 from model import IrisNet
+import time
```
Comment (deliahu) [inline test/apis/pytorch/iris-classifier/predictor.py:8]: can be deleted

File: test/utils/throughput_test.py [hunk 15]
```diff
@@ -22,15 +22,15 @@
     type=int,
     default=1,
     show_default=True,
-    help="Number of processes for prediction requests.",
+    help="Number of processes for compute requests.",
```
Comment (deliahu) [inline test/utils/throughput_test.py:25]: nit: delete "compute" (and below)

**Review by deliahu — APPROVED** [review]

File: pkg/lib/k8s/service.go [hunk 16]
```diff
@@ -158,16 +158,7 @@ func (c *Client) ListServicesWithLabelKeys(labelKeys ...string) ([]kcore.Service
 }
 
 func (c *Client) InternalServiceEndpoint(serviceName string, portNumber int32) (string, error) {
```
Comment (deliahu) [inline pkg/lib/k8s/service.go:160]: this doesn't have to return an error

File: pkg/operator/resources/realtimeapi/api.go [hunk 17]
```diff
@@ -487,15 +487,15 @@ func getAPIModelSummary(apiEndpoint string) (*schema.APIModelSummary, error) {
...
+	var apiModelSummary schema.PythonModelSummary
 	err = json.DecodeWithNumber(response, &apiModelSummary)
 	if err != nil {
 		return nil, errors.Wrap(err, "unable to parse api summary response")
 	}
 	return &apiModelSummary, nil
 }
 
-func getAPITFLiveReloadingSummary(apiEndpoint string) (*schema.APITFLiveReloadingSummary, error) {
+func getAPITFLiveReloadingSummary(apiEndpoint string) (*schema.TFLiveReloadingSummary, error) {
```
Comment (deliahu) [inline pkg/operator/resources/realtimeapi/api.go:498]: nit: can remove "API" from this function name (and variables in this function)

File: pkg/operator/resources/realtimeapi/api.go [hunk 18]
```diff
@@ -476,7 +476,7 @@ func GetModelsMetadata(status *status.Status, handler *userconfig.Handler, apiEn
 	return nil, nil, nil
 }
 
-func getAPIModelSummary(apiEndpoint string) (*schema.APIModelSummary, error) {
+func getAPIModelSummary(apiEndpoint string) (*schema.PythonModelSummary, error) {
```
Comment (deliahu) [inline pkg/operator/resources/realtimeapi/api.go:479]: `getAPIModelSummary` -> `getPythonModelSummary`

File: pkg/operator/resources/resources.go [hunk 19]
```diff
@@ -621,8 +621,8 @@ func GetAPI(apiName string) ([]schema.APIResponse, error) {
 			return apiResponse, nil
 		}
 
-		apiResponse[0].RealtimeModelMetadata.APITFModelSummary = apiTFModelSummary
-		apiResponse[0].RealtimeModelMetadata.APIPythonModelSummary = apiPythonModelSummary
+		apiResponse[0].RealtimeModelMetadata.TFModelSummary = apiTFModelSummary
+		apiResponse[0].RealtimeModelMetadata.PythonModelSummary = apiPythonModelSummary
```
Comment (deliahu) [inline pkg/operator/resources/resources.go:625]: remove "api" on these lines

**Review by deliahu — APPROVED** [review]
